File: src/js/4_BehavioralPatterns/11_Visitor/classic.js
class NumberExpression
{
  constructor(value)
  {
    this.value = value;
  }

  accept(visitor)
  {
    visitor.visitNumber(this);
  }
}

class AdditionExpression
{
  constructor(left, right)
  {
    this.left = left;
    this.right = right;
  }

  accept(visitor)
  {
    visitor.visitAddition(this);
  }
}

class Visitor
{
  constructor()
  {
    this.buffer = [];
  }
}

class ExpressionPrinter extends Visitor
{
  constructor()
  {
    super();
  }

  visitNumber(e)
  {
    this.buffer.push(e.value);
  }

  visitAddition(e)
  {
    this.buffer.push('(');
    e.left.accept(this);
    this.buffer.push('+');
    e.right.accept(this);
    this.buffer.push(')');
  }

  toString()
  {
    return this.buffer.join('');
  }
}

class ExpressionCalculator
{
  // this visitor is stateful which can lead to problems
  constructor()
  {
    this.result = 0;
  }

  visitNumber(e)
  {
    this.result = e.value;
  }

  visitAddition(e)
  {
    e.left.accept(this);
    let temp = this.result;
    e.right.accept(this);
    this.result += temp;
  }
}

let e = new AdditionExpression(
  new NumberExpression(1),
  new AdditionExpression(
    new NumberExpression(2),
    new NumberExpression(3)
  )
);

var ep = new ExpressionPrinter();
ep.visitAddition(e);

var ec = new ExpressionCalculator();
ec.visitAddition(e);

console.log(
  `${ep.toString()} = ${ec.result}`
);

File: src/go/4_BehavioralPatterns/11_Visitor/classic.go
package visitor

import (
  "fmt"
  "strings"
)

type ExpressionVisitor interface {
  VisitDoubleExpression(de *DoubleExpression)
  VisitAdditionExpression(ae *AdditionExpression)
}

type Expression interface {
  Accept(ev ExpressionVisitor)
}

type DoubleExpression struct {
  value float64
}

func (d *DoubleExpression) Accept(ev ExpressionVisitor) {
  ev.VisitDoubleExpression(d)
}

type AdditionExpression struct {
  left, right Expression
}

func (a *AdditionExpression) Accept(ev ExpressionVisitor) {
  ev.VisitAdditionExpression(a)
}

type ExpressionPrinter struct {
  sb strings.Builder
}

func (e *ExpressionPrinter) VisitDoubleExpression(de *DoubleExpression) {
  e.sb.WriteString(fmt.Sprintf("%g", de.value))
}

func (e *ExpressionPrinter) VisitAdditionExpression(ae *AdditionExpression) {
  e.sb.WriteString("(")
  ae.left.Accept(e)
  e.sb.WriteString("+")
  ae.right.Accept(e)
  e.sb.WriteString(")")
}

func NewExpressionPrinter() *ExpressionPrinter {
  return &ExpressionPrinter{strings.Builder{}}
}

func (e *ExpressionPrinter) String() string {
  return e.sb.String()
}

func main() {
  // 1+(2+3)
  e := &AdditionExpression{
    &DoubleExpression{1},
    &AdditionExpression{
      left:  &DoubleExpression{2},
      right: &DoubleExpression{3},
    },
  }
  ep := NewExpressionPrinter()
  ep.VisitAdditionExpression(e)
  fmt.Println(ep.String())
}